internal/controllers/machineset/machineset_controller.go

```diff
@@ -883,8 +888,7 @@ func (r *Reconciler) updateStatus(ctx context.Context, cluster *clusterv1.Cluste
 
 		node, err := r.getMachineNode(ctx, cluster, machine)
 		if err != nil && machine.GetDeletionTimestamp().IsZero() {
-			log.Error(err, "Unable to retrieve Node status", "node", klog.KObj(node))
-			continue
+			return errors.Wrapf(err, "unable to retrieve the status of Node %s", klog.KObj(node))
```

Comment:
I think this doesn't address:

> For MachineSet specifically: It definitely solves the "happy path", but it makes the **"unhappy path" worse by freezing the status indefinitely**. Without the changes the ready and available replicas were dropping to 0 when the communication broke down. Probably a matter of preference if in case of the communication breakdown we prefer to keep ready and available replicas as is or drop them to 0. **But another side effect is also that we don't update ObservedGeneration anymore, which definitely seems wrong (similar for the conditions)**

Reply:
Thanks @sbueringer.

> "unhappy path" worse by freezing the status indefinitely

This PR does not update any fields of MS.Status when unable to get workload cluster or machine Node due to ErrClusterLocked or any other errors. Because the ErrClusterLocked error can be recovered soon after reconciling again, and the error that cannot get machine Node (e.g. network issue, or apiserver unavailable temporarily) should also be recovered soon after reconciling again.  Are there other "unhappy path" you meant?

> But another side effect is also that we don't update ObservedGeneration anymore, which definitely seems wrong (similar for the conditions).

Since this PR returns error at line 891, so it won't update ms.Status at line 904 and newStatus.ObservedGeneration at line 924. This is as expected.

![图片](https://github.com/kubernetes-sigs/cluster-api/assets/10006/dba0db48-77e5-4d88-b766-e6b114bd6120)
![图片](https://github.com/kubernetes-sigs/cluster-api/assets/10006/408def26-32ba-4f19-9c70-028fe2c35c5c)

Reply:
> and the error that cannot get machine Node (e.g. network issue, or apiserver unavailable temporarily) should also be recovered soon after reconciling again. Are there other "unhappy path" you meant?

Yup, apiserver down for a longer period of time / indefinitely

Reply:
In case the apiserver goes down for a longer period of time or indefinitely, KCP controller should catch it and handle correctly.

Reply:
KCP won't always be able to recover (if that's what you meant)

Reply:
I meant KCP controller should handle the case that the apiserver goes down for a longer period of time or indefinitely. Even if MS controller detects it, it can't handle it except updating MS status replicas.

Reply:
Yup agree. MS controller can't fix it. What we're saying is just that the MS controller shouldn't stop reconciling all MS.status fields because of that

Reply:
I think in this case reconciling other MS.status fields does not help. Instead let KCP controller handles it, e.g. update its status fields, and notifies the consumers.

Reply:
BTW, a KCP status related PR was merged recently [📖 Improve docs for KCP.status.ready & cluster.status.ControlPlaneReady by pravarag · Pull Request #10355 · kubernetes-sigs/cluster-api](https://github.com/kubernetes-sigs/cluster-api/pull/10355): When I was verifying the failure scenario, I manually deleted the CP Machine, and the apiserver was unreachable, KCP.status.readyReplicas is 0 but KCP.status.ready is always in true.

Reply:
hi @fabriziopandini would you mind take a look at this refined patch again ? Thanks.

Reply:
Hi @sbueringer could you kindly help take further action (more comments or suggestions) on this PR please?  Looks like @fabriziopandini is too busy recently to share more idea on this new PR.

Reply:
Q: is this the same issue for which we suggested a solution in https://github.com/kubernetes-sigs/cluster-api/pull/10229#discussion_r1557829121
Why are we deviating from the suggested solution?

Reply:
Hi @fabriziopandini as discussed with @sbueringer in current thread,  IMHO this PR provides a simple patch to solve the MS status update issue described in https://github.com/kubernetes-sigs/cluster-api/issues/10195, and there would be no need to introduce the suggested handling code (provided by @sbueringer and IMHO a little complex)
1) I had replied the my thought in https://github.com/kubernetes-sigs/cluster-api/pull/10229#discussion_r1560899466 and removed the KCP status handling code as you suggested.
2) In case the apiserver goes down for a longer period of time or indefinitely, the MS status won't be updated, but KCP controller should catch it and handle correctly.  In this case updating MS status won't help.

BTW I was on vacation last week so sorry for the late reply. Thanks!

Reply:
I will look into it as soon as possible, but we already invested a lot of time in trying to find an acceptable way forward.

Reply:
I tried to express above that it's not acceptable that the MS conditions are permanently not updated anymore when a workload cluster stays unreachable

Reply:
As I described that 
2) in case the apiserver goes down for a longer period of time or indefinitely, the MS status won't be updated, but KCP controller should catch it and handle correctly. In this case updating MS status won't help.  
3) Even if we want to update MS status in this special case,  it would be not good to set th MS ready replicas to 0, instead IMHO a new status UnknownReplicas should be added (but this changes the current API definition and is not a trivial change).